modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Unfair.java

```diff
@@ -0,0 +1,764 @@
+
+            acquireProcessor = new AcquireUnfairProcessor(nodeId);
+            releaseProcessor = new ReleaseUnfairProcessor(nodeId);
+            lockIfFreeProcessor = new AcquireIfFreeUnfairProcessor(nodeId);
+            lockOrRemoveProcessor = new AcquireOrRemoveUnfairProcessor(nodeId);
+
+            acquireListener = new IgniteInClosure<IgniteInternalFuture<EntryProcessorResult<Boolean>>>() {
+                @Override public void apply(IgniteInternalFuture<EntryProcessorResult<Boolean>> future) {
+                    try {
+                        EntryProcessorResult<Boolean> result = future.get();
+
+                        if (result.get())
+                            latch.release();
+                    }
+                    catch (IgniteCheckedException e) {
+                        if (log.isDebugEnabled())
+                            log.debug("Acquire invoke has failed: " + e);
+
+                        latch.fail(U.convertException(e));
+                    }
+                }
+            };
+
+            releaseListener = new IgniteInClosure<IgniteInternalFuture<EntryProcessorResult<UUID>>>() {
+                @Override public void apply(IgniteInternalFuture<EntryProcessorResult<UUID>> future) {
+                    try {
+                        EntryProcessorResult<UUID> result = future.get();
+
+                        // invokeAsync return null if EntryProcessor return null too.
+                        if (result != null) {
+                            UUID nextNode = result.get();
+
+                            if (nextNode != null) {
+                                if (nodeId.equals(nextNode))
+                                    latch.release();
+                                else {
+                                    ctx.io().send(nextNode,
+                                        new GridCacheLockImpl2Unfair.ReleasedMessage(ctx.cacheId(), key.name()), P2P_POOL);
+                                }
+                            }
+                        }
+                    }
+                    catch (IgniteCheckedException e) {
+                        if (log.isDebugEnabled())
+                            log.debug("Release message sending has failed: " + e);
+                    }
+                }
+            };
+        }
+
+        /** */
+        private IgniteInternalFuture<EntryProcessorResult<Boolean>> tryAcquireOrAdd() {
+            return lockView.invokeAsync(key, acquireProcessor);
+        }
+
+        /**
+         * Acquires the lock only if it is not held by another node at the time
+         * of invocation.
+         *
+         * @return {@code true} if the lock was free and was acquired by the
+         *          current node, or the lock was already held by the current
+         *          node; and {@code false} otherwise.
+         */
+        private boolean tryAcquire() {
+            try {
+                return lockView.invoke(key, lockIfFreeProcessor).get();
```

Comment:
idea says it could produce npe) may be just add assert

Reply:
No, it can't